fw/apm.c

```diff
@@ -639,17 +628,42 @@ typedef struct {
 #define TFW_APM_DFLT_REQ	"\"GET / HTTTP/1.0\r\n\r\n\""
 #define TFW_APM_DFLT_URL	"\"/\""
 
+/*
+ * APM Data structure.
+ *
+ * Note that the organization of the supporting data heavily depends
+ * on the fact that there's only one party that does the calculation
+ * of percentiles - the function that runs periodically on timer.
+ * If there are several different parties that do the calculation,
+ * then the data may need to be organized differently.
+ *
+ * @rbuf	- The ring buffer for the specified time window.
+ * @rbctl	- The control data helpful in taking optimizations.
+ * @stats	- The latest percentiles.
+ * @ubuf	- The buffer that holds data for updates, per CPU.
+ * @timer	- The periodic timer handle.
+ * @flags	- The atomic flags (see below).
+ */
 typedef struct {
 	TfwApmRBuf		rbuf;
 	TfwApmRBCtl		rbctl;
 	TfwApmStats		stats;
 	TfwApmUBuf __percpu	*ubuf;
-	TfwServer		*srv;
 	struct timer_list	timer;
 	unsigned long		flags;
-	TfwApmHMCtl		hmctl;
 } TfwApmData;
 
+/*
+ * Order of the members is important for the memory allocation, @data must be
+ * at the end.
+ */
+typedef struct {
+	TfwApmHMCtl		hmctl;
+	atomic64_t		hits;
+	atomic64_t		misses;
+	TfwApmData		data;
+} TfwApmRef;
```

Comment:
Why do we need the new data structure, why can't (shouldn't) we keep the cache misses and hits in `TfwApmData`?

Reply:
We need TfwApmData separately, without heath monitor, for the global statistics (`tfw_apm_global_data`). Hits and misses also don't required here.